### PR TITLE
Implement deallocation events

### PR DESCRIPTION
**Pull request by interwq**
Make the event module to accept two event types, and pass around the event context.  Use bytes-based events to trigger tcache GC on deallocation, and get rid of the tcache ticker.

### CODE REVIEW

**Comment by interwq** [comment]
This speeds up the deallocation common path.
Results from `test/stress/microbench`:

Before this PR, based on #1749 (to properly measure `sdallocx`)

    1000000000 iterations, free=11934090us, sdallocx=7581057us, ratio=1:1.574

With this PR

    1000000000 iterations, free=11707088us, sdallocx=7112053us, ratio=1:1.646

So the change in free is small (a couple of %, though still consistent); but sized free shows pretty nice win. The round trip `malloc(); sdallocx()` time reduced by ~7%.

I do wonder how the tcache GC is affected. Benchmarking in production next; let's see if the numbers on paper convert to real wins this time.

File: src/thread_event.c [hunk 1]
```diff
@@ -267,28 +306,30 @@ thread_event_trigger_batch_update(tsd_t *tsd, uint64_t accumbytes,
 }
 
 void
-thread_event_trigger(tsd_t *tsd, bool delay_event) {
+thread_event_trigger(tsd_t *tsd, event_ctx_t *ctx, bool is_alloc,
+    bool delay_event) {
```
Comment (yinan1048576) [inline src/thread_event.c]: Is `is_alloc` always equal to `ctx->is_alloc`?

File: include/jemalloc/internal/thread_event.h [hunk 2]
```diff
@@ -4,42 +4,52 @@
...
-    E(tcache_gc,	(TCACHE_GC_INCR_BYTES > 0))			\
-    E(prof_sample,	(config_prof && opt_prof))	    		\
-    E(stats_interval,	(opt_stats_interval >= 0))
+    E(tcache_gc,	(TCACHE_GC_INCR_BYTES > 0), true)		\
+    E(prof_sample,	(config_prof && opt_prof), true)	    	\
+    E(stats_interval,	(opt_stats_interval >= 0), true)	    	\
+    E(tcache_gc_dalloc,	(TCACHE_GC_INCR_BYTES > 0), false)
 
-#define E(event, condition)						\
+#define E(event, ignored, ignored1)					\
```
[the remark this answers is not in the document]
Comment (interwq) [reply include/jemalloc/internal/thread_event.h]: Good point. Will update.

File: include/jemalloc/internal/tsd.h [hunk 3]
```diff
@@ -81,10 +81,14 @@ typedef void (*test_callback_t)(int *);
     O(thread_allocated,		uint64_t,		uint64_t)	\
     O(thread_allocated_next_event_fast,	uint64_t,	uint64_t)	\
     O(thread_deallocated,	uint64_t,		uint64_t)	\
+    O(thread_deallocated_next_event_fast, uint64_t,	uint64_t)	\
```
[the remark this answers is not in the document]
Comment (interwq) [reply include/jemalloc/internal/tsd.h]: How about I update that part in a separate PR? Have a refactoring coming up soon.

File: src/jemalloc.c [hunk 4]
```diff
@@ -2706,6 +2705,13 @@ bool free_fastpath(void *ptr, size_t size, bool size_hint) {
 		}
 		assert(szind != SC_NSIZES);
 	} else {
+		/*
+		 * tsd_fast check is folded into the branch testing
+		 * fast_threshold (set to 0 when !tsd_fast).
+		 */
+		if (unlikely(!tsd)) {
+			return false;
+		}
```
[the remark this answers is not in the document]
Comment (interwq) [reply src/jemalloc.c]: I should beef up the comments here. With size_hint, we don't have to do rtree_lookup (which requires tsd to be initialized). If the path below can tolerate uninitialized tsd, the tsd_fast check is done through the next_event_fast threshold.

Does that make sense?

File: src/thread_event.c (hunk 1 above)
Comment (interwq) [reply src/thread_event.c]: Yep good point. Will update.

File: src/jemalloc.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (yinan1048576) [reply src/jemalloc.c]: Yes, it makes sense now.

File: include/jemalloc/internal/tsd.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (yinan1048576) [reply include/jemalloc/internal/tsd.h]: Sounds good.

**Comment by interwq** [comment]
@yinan1048576 added a separate commit addressing the feedback. I'll merge it back to the main commit if it looks good.

**Comment by yinan1048576** [comment]
LGTM.

**Comment by interwq** [comment]
Production testing done. In the main binaries benchmarked, deallocation-related (i.e. stack trace containing `.*sdallocx.*`) cost reduced by 5-6%, notably the ticker related cost on free_fastpath is gone (arena decay still ticks, ticker-related reduced by ~90%).

Disassembly of updated sdallocx (cost distribution from stress benchmark):

       │    000000000001d370 <sdallocx>:                                                                                                                                                                                                                                                                                                             ▒
       │    sdallocx():                                                                                                                                                                                                                                                                                                                              ▒
       │    JEMALLOC_EXPORT void JEMALLOC_NOTHROW                                                                                                                                                                                                                                                                                                    ▒
       │    je_sdallocx(void *ptr, size_t size, int flags) {                                                                                                                                                                                                                                                                                         ▒
       │            LOG("core.sdallocx.entry", "ptr: %p, size: %zu, flags: %d", ptr,                                                                                                                                                                                                                                                                 ▒
       │                    size, flags);                                                                                                                                                                                                                                                                                                            ▒
       │                                                                                                                                                                                                                                                                                                                                             ▒
       │            if (flags != 0 || !free_fastpath(ptr, size, true)) {                                                                                                                                                                                                                                                                             ▒
    0.10 │      test   %edx,%edx                                                                                                                                                                                                                                                                                                                       ▒
       │    . jne    90                                                                                                                                                                                                                                                                                                                              ▒
       │    free_fastpath():                                                                                                                                                                                                                                                                                                                         ▒
       │                    if (unlikely(size > SC_LOOKUP_MAXCLASS ||                                                                                                                                                                                                                                                                                ▒
       │      cmp    $0x1000,%rsi                                                                                                                                                                                                                                                                                                                    ▒
       │    . ja     90                                                                                                                                                                                                                                                                                                                              ▒
    0.01 │      test   $0xfff,%edi                                                                                                                                                                                                                                                                                                                     ▒
       │    . je     90                                                                                                                                                                                                                                                                                                                              ▒
       │    sz_size2index_lookup_impl():                                                                                                                                                                                                                                                                                                             ▒
       │    }                                                                                                                                                                                                                                                                                                                                        ▒
       │                                                                                                                                                                                                                                                                                                                                             ▒
       │    JEMALLOC_ALWAYS_INLINE szind_t                                                                                                                                                                                                                                                                                                           ▒
       │    sz_size2index_lookup_impl(size_t size) {                                                                                                                                                                                                                                                                                                 ▒
       │            assert(size <= SC_LOOKUP_MAXCLASS);                                                                                                                                                                                                                                                                                              ▒
       │            return sz_size2index_tab[(size + (ZU(1) << SC_LG_TINY_MIN) - 1)                                                                                                                                                                                                                                                                  ▒
    9.14 │      lea    0x7(%rsi),%rax                                                                                                                                                                                                                                                                                                                  ▒
       │    sz_index2size_lookup_impl():                                                                                                                                                                                                                                                                                                             ▒
       │            }                                                                                                                                                                                                                                                                                                                                ▒
       │    }                                                                                                                                                                                                                                                                                                                                        ▒
       │                                                                                                                                                                                                                                                                                                                                             ▒
       │    JEMALLOC_ALWAYS_INLINE size_t                                                                                                                                                                                                                                                                                                            ▒
       │    sz_index2size_lookup_impl(szind_t index) {                                                                                                                                                                                                                                                                                               ▒
       │            return sz_index2size_tab[index];                                                                                                                                                                                                                                                                                                 ▒
    0.06 │      lea    je_sz_size2index_tab,%rcx                                                                                                                                                                                                                                                                                                       ▒
       │    thread_event_free_fastpath_ctx():                                                                                                                                                                                                                                                                                                        ▒
       │            if (!size_hint) {                                                                                                                                                                                                                                                                                                                ▒
       │                    *deallocated = tsd_thread_deallocated_get(tsd);                                                                                                                                                                                                                                                                          ▒
       │                    *threshold = tsd_thread_deallocated_next_event_fast_get(tsd);                                                                                                                                                                                                                                                            ▒
       │            } else {                                                                                                                                                                                                                                                                                                                         ▒
       │                    /* Unsafe getters since this may happen before tsd_init. */                                                                                                                                                                                                                                                              ▒
       │                    *deallocated = *tsd_thread_deallocatedp_get_unsafe(tsd);                                                                                                                                                                                                                                                                 ▒
    0.01 │      mov    .got+0x68,%r11                                                                                                                                                                                                                                                                                                                  ▒
       │    free_fastpath():                                                                                                                                                                                                                                                                                                                         ▒
       │            uint64_t deallocated_after = deallocated + usize;                                                                                                                                                                                                                                                                                ▒
       │      lea    je_sz_index2size_tab,%r9                                                                                                                                                                                                                                                                                                        ▒
       │    sz_size2index_lookup_impl():                                                                                                                                                                                                                                                                                                             ▒
       │                >> SC_LG_TINY_MIN];                                                                                                                                                                                                                                                                                                          ▒
    25.88 │      shr    $0x3,%rax                                                                                                                                                                                                                                                                                                                       ▒
       │    sz_index2size_lookup_impl():                                                                                                                                                                                                                                                                                                             ▒
       │            return sz_index2size_tab[index];                                                                                                                                                                                                                                                                                                 ▒
    0.04 │      movzbl (%rcx,%rax,1),%r8d                                                                                                                                                                                                                                                                                                              ▒
       │    free_fastpath():                                                                                                                                                                                                                                                                                                                         ▒
    0.01 │      mov    (%r9,%r8,8),%rax                                                                                                                                                                                                                                                                                                                ▒
    0.02 │      add    %fs:0x18(%r11),%rax                                                                                                                                                                                                                                                                                                             ▒
       │            if (unlikely(deallocated_after >= threshold)) {                                                                                                                                                                                                                                                                                  ▒
    14.12 │      cmp    %rax,%fs:0x20(%r11)                                                                                                                                                                                                                                                                                                             ▒
       │    . jbe    90                                                                                                                                                                                                                                                                                                                              ▒
       │            cache_bin_t *bin = tcache_small_bin_get(tcache, szind);                                                                                                                                                                                                                                                                          ▒
    0.07 │      mov    %r11,%rcx                                                                                                                                                                                                                                                                                                                       ▒
       │      add    %fs:0x0,%rcx                                                                                                                                                                                                                                                                                                                    ▒
       │      lea    (%r8,%r8,2),%r10                                                                                                                                                                                                                                                                                                                ▒
    14.07 │      lea    (%rcx,%r10,8),%r8                                                                                                                                                                                                                                                                                                               ▒
       │    cache_bin_dalloc_easy():                                                                                                                                                                                                                                                                                                                 ▒
       │                                                                                                                                                                                                                                                                                                                                             ▒
       │    #undef INVALID_SZIND                                                                                                                                                                                                                                                                                                                     ▒
       │                                                                                                                                                                                                                                                                                                                                             ▒
       │    JEMALLOC_ALWAYS_INLINE bool                                                                                                                                                                                                                                                                                                              ▒
       │    cache_bin_dalloc_easy(cache_bin_t *bin, void *ptr) {                                                                                                                                                                                                                                                                                     ▒
       │            if (unlikely(bin->cur_ptr.lowbits == bin->full_position)) {                                                                                                                                                                                                                                                                      ▒
    0.55 │      mov    0x25c(%r8),%r9d                                                                                                                                                                                                                                                                                                                 ▒
    0.49 │      cmp    %r9d,0x248(%r8)                                                                                                                                                                                                                                                                                                                 ▒
       │    . je     90                                                                                                                                                                                                                                                                                                                              ▒
       │                    return false;                                                                                                                                                                                                                                                                                                            ▒
       │            }                                                                                                                                                                                                                                                                                                                                ▒
       │                                                                                                                                                                                                                                                                                                                                             ▒
       │            *(--bin->cur_ptr.ptr) = ptr;                                                                                                                                                                                                                                                                                                     ▒
    0.54 │      mov    0x248(%r8),%rdx                                                                                                                                                                                                                                                                                                                 ▒
    18.39 │      lea    -0x8(%rdx),%rsi                                                                                                                                                                                                                                                                                                                 ▒
    1.10 │      mov    %rsi,0x248(%r8)                                                                                                                                                                                                                                                                                                                 ▒
    7.66 │      mov    %rdi,-0x8(%rdx)                                                                                                                                                                                                                                                                                                                 ▒
       │    free_fastpath():                                                                                                                                                                                                                                                                                                                         ▒
       │            *tsd_thread_deallocatedp_get(tsd) = deallocated_after;                                                                                                                                                                                                                                                                           ▒
    0.04 │      mov    %rax,%fs:0x18(%r11)                                                                                                                                                                                                                                                                                                             ▒
    7.70 │    , retq                                                                                                                                                                                                                                                                                                                                   ▒
       │      nop                                                                                                                                                                                                                                                                                                                                    ▒
       │    sdallocx():                                                                                                                                                                                                                                                                                                                              ▒
       │                    sdallocx_default(ptr, size, flags);                                                                                                                                                                                                                                                                                      ▒
       │90: + jmpq   1c240 <je_sdallocx_default>

**Pull request merged**